cpp/source/deal.II/new_supercapacitor.templates.h

```diff
@@ -0,0 +1,358 @@
+/* Copyright (c) 2016, the Cap authors.
+ *
+ * This file is subject to the Modified BSD License and may not be distributed
+ * without copyright and license information. Please refer to the file LICENSE
+ * for the text and further information on this license.
+ */
+
+#ifndef CAP_DEAL_II_NEW_SUPERCAPACITOR_TEMPLATES_H
+#define CAP_DEAL_II_NEW_SUPERCAPACITOR_TEMPLATES_H
+
+#include <cap/deal.II/new_supercapacitor.h>
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/dofs/dof_renumbering.h>
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_tools.h>
+#include <deal.II/numerics/matrix_tools.h>
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/lac/precondition.h>
+#include <deal.II/lac/solver_cg.h>
+#include <boost/format.hpp>
+#include <boost/foreach.hpp>
+#include <boost/property_tree/ptree.hpp>
+#include <boost/test/floating_point_comparison.hpp>
+#include <tuple>
+#include <fstream>
+
+namespace cap
+{
+template <int dim>
+New_SuperCapacitor<dim>::New_SuperCapacitor(
+    boost::mpi::communicator const &comm,
+    boost::property_tree::ptree const &ptree)
+    : EnergyStorageDevice(comm)
+{
+  // get database
+  boost::property_tree::ptree const &database = ptree;
+
+  // build triangulation
+  std::shared_ptr<boost::property_tree::ptree> geometry_database =
+      std::make_shared<boost::property_tree::ptree>(
+          database.get_child("geometry"));
+  geometry =
+      std::make_shared<cap::SuperCapacitorGeometry<dim>>(geometry_database);
+  std::shared_ptr<dealii::Triangulation<dim> const> triangulation =
+      geometry->get_triangulation();
+
+  // get data tolerance and maximum number of iterations for the CG solver
+  boost::property_tree::ptree const &solver_database =
+      database.get_child("solver");
+  max_iter      = solver_database.get<unsigned int>("max_iter", 1000);
+  rel_tolerance = solver_database.get<double>("rel_tolerance", 1e-12);
+  abs_tolerance = solver_database.get<double>("abs_tolerance", 1e-12);
+  verbose_lvl   = solver_database.get<unsigned int>("verbosity", 0);
```

Comment:
all these are ``solver`` parameters. It advocates for a solver object in the future.